active_discussion/representation.md

```diff
@@ -1,14 +1,125 @@
-  - e.g., `&T` and `Option<&T>` are both pointer sized
-  - size of `extern fn` etc (at least on some platforms)?
-  - For which `T` is `None` represented as a "null pointer" etc?
-    - (Which "niche" optimizations can we rely on)
+- What guarantees does Rust make regarding the layout of data structures?
+- What invariants does the compiler require from the various Rust types?
+  - the "validity invariant", as defined in [Ralf's blog post][bp]
+
+NB. The discussion is **not** meant to discuss the "safety invariant"
+from [Ralf's blog post][bp], as that can be handled later.
+
+[bp]: https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html
+
+### Layout of data structures
+
+In general, Rust makes few guarantees about the memory layout of your
+structures. For example, by default, the compiler has the freedom to
+rearrange the field order of your structures for more efficiency (as
+of this writing, we try to minimize the overall size of your
+structure, but this is the sort of detail that can easily change). For
+safe code, of course, any rearrangements "just work" transparently.
+
+If, however, you need to write unsafe code, you may wish to have a
+fixed data structure layout. In that case, there are ways to specify
+and control how an individual struct will be laid out -- notably with
+`#[repr]` annotations. One purpose of this section, then, is to layout
+what sorts of guarantees we offer when it comes to layout, and also
+what effect the various `#[repr]` annotations have.
+
+### Validity invariant
+
+The "validity invariant" for each type defines what must hold whenever
+a value of this type is considered to be initialized. The compiler expects
+the validity invariant to hold **at all times** and is thus allowed to use
+these invariants to (e.g.) affect the layout of data structures or do other
+optimizations.
+
+Therefore, the validity invariant must **at minimum** justify all the
+layout optimizations that the compiler does. We may want a stronger
+invariant, however, so as to leave room for future optimization.
+
+As an example, a value of `&T` type can never be null -- therefore,
+`Option<&T>` can use null to represent `None`.
+
+## Goals
+
+- Define what we guarantee about the layout of various types
+  and the effect of `#[repr]` annotations.
+- Define the **validity requirements** of various types that unsafe
+  programmers must uphold at all times.
+    - Also examine when/how we could dynamically check these requirements.
+- Uncover the sorts of constraints that we may wish to satisfy in the
+  future.
+
+## Some interesting examples and questions
+
+- `&T` where `T: Sized`
+  - This is **guaranteed** to be a non-null pointer
+- `Option<&T>` where `T: Sized`
+  - This is **guaranteed** to be a nullable pointer
+- `Option<extern "C" fn()>`
+- `usize`
+  - Platform dependent size, but guaranteed to be able to store a pointer?
+  - Also an array length?
```

Comment:
C++ says that `usize` is an unsigned integer type that can store the maximum size (as returned by `mem::size_of<T>`/`size_of_val`/etc.) of a theoretically possible object of any type (including arrays).

Reply:
Also, it might be worth it to talk about `isize` here. C++ says: If an array is so large (greater than `isize::max_value()` elements, but less than `usize::max_value()` bytes), that the difference between two pointers may not be representable as `isize`, the result of subtracting two such pointers is undefined.

Reply:
This reminds me that Rust currently has a rule that the maximum size of any value is representable with `isize`, I believe. I'm not sure how "baked in" that rule is by now -- probably "fairly" -- but it'd be good to document (and, if we don't like it, think seriously about what it takes to change it). I'll add a note to that effect.

Reply:
That sounds reasonable and should simplify the definitions of both `usize` and `isize`.

Reply:
@nikomatsakis I think it is also worth discussing whether we want to make `0_usize`/`0_isize` `as *_ T` be equal to `ptr::null/null_mut` or not.

AFAIK in C and C++ at least `NULL` isn't necessary equal to the address at `0`, but rather a platform-specific address. Currently, we do not promise this in Rust, but we do many optimizations like `Option<&T>` which are all hardcoded to `0` being a special bitpattern. The question is, should people only be able to use `.is_null` and `ptr::null()` to construct and test for null pointers, or should they also be able to cast a specific integer into a pointer that creates a null pointer, and should this integer _always_ be 0, or be `isize/usize::consts::NULL` or something like that?

This might affect some of the wording we use (e.g. non-zero optimization, ...) everywhere.

Reply:
added various notes about these things :)

Reply:
@gnzlbg AFAIK no, while the *constant expression* `0` turns into the NULL pointer when cast to a pointer, a runtime integer value cast to `0` does not necessarily have to turn into the NULL pointer when cast at run time (int<->ptr casts are implementation-defined anyway IIRC).

Reply:
@rkruppe indeed, the zero is null is only so for constant expressions .

Reply:
That's C++, though, right? For Rust we can just say that a pointer is NULL iff its bits are all 0...

Reply:
Yes, we can say that, and we would exclude some (possibly hypothetical) targets/implementations by doing that. That's what needs to be discussed.